nengo/builder/tests/test_transforms.py

```diff
@@ -109,9 +110,213 @@ def test_convinc_2d(
+        signals_tr[xt][:] = C[k].reshape(conv_transpose.input_shape.shape)
+        signals_tr[yt][:] = 0
+        step_trans()
+        AtCt.append(signals_tr[yt].copy().ravel())
+
+    AtCt = np.array(AtCt)
+    conv_transpose_output = AtCt.dot(x_flat)
+    assert conv_transpose_output.shape == conv_output.shape
+
+    # Check whether the outputs match. If they don't, print/plot some diagnostic output.
+    success = allclose(conv_transpose_output, conv_output)
+    if not success:
+        # transpose convolution should act as the derivative for forward convolution,
+        # compute and compare to the derivative to see exactly where the difference is
+        signals = {sig: np.array(sig.initial_value) for sig in (x, w, y)}
+        step_conv = ConvInc(w, x, y, conv).make_step(signals, None, None)
+        step_conv()
+
+        dy_dx = []
+        for j in range(conv.input_shape.size):
+            xj = np.zeros(conv.input_shape.shape)
+            xj.ravel()[j] = 1
+
+            signals[x][:] = xj
+            signals[y][:] = 0
+            step_conv()
+
+            dy_dx.append(signals[y].copy().ravel())
+
+        dy_dx = np.column_stack(dy_dx)
+        assert dy_dx.shape == (conv.output_shape.size, conv.input_shape.size)
+
+        signals_tr = {sig: np.array(sig.initial_value) for sig in (xt, wt, yt)}
+        step_trans = ConvTransposeInc(wt, xt, yt, conv_transpose).make_step(
+            signals_tr, None, None
+        )
+
+        for i in range(conv.output_shape.size):
+            xi = np.zeros(conv.output_shape.shape)
+            xi.ravel()[i] = 1
+
+            signals_tr[xt][:] = xi
+            signals_tr[yt][:] = 0
+            step_trans()
+
+            actual = signals_tr[yt]
+            expected = dy_dx[i, :].reshape(conv.input_shape.shape)
+
+            assert actual.shape == expected.shape
+            if not np.allclose(actual, expected):
+                for k in range(in_channels):
+                    actual_k = actual[:, :, k] if channels_last else actual[k]
+                    expected_k = expected[:, :, k] if channels_last else expected[k]
+
+                    if not np.allclose(actual, expected):
+                        plt.subplot(121)
+                        plt.imshow(
+                            expected_k, vmin=expected_k.min(), vmax=expected_k.max()
+                        )
+                        plt.title(f"expected out={i} ch={k}")
+
+                        plt.subplot(122)
+                        plt.imshow(
+                            actual_k, vmin=expected_k.min(), vmax=expected_k.max()
+                        )
+                        plt.title(f"actual out={i} ch={k}")
+
+                        assert success, f"Output {i} channel {k} differs"
+
+    assert success
+
+
+@pytest.mark.parametrize("x_size", (1, 2, 3, 4))
+@pytest.mark.parametrize("k_size", (1, 2, 3, 4))
+@pytest.mark.parametrize("stride", (1, 2, 3, 4))
+@pytest.mark.parametrize("padding", ("same", "valid"))
```

Comment:
The other thing we can do if the concern is really noise in the test output, rather than the speed, is do the loop inside the test rather than with `parametrize`.

The reason I usually do [1, 2, 3, 4] is that 1 and 2 are fairly special (in that they're smaller than the typical values used for a lot of these parameters), and 3 and 4 give one even and one odd value in a more "normal" range (though you could also make an argument to use e.g. 6 and 9 instead of 3 and 4, but you'll still have 4 values for each one).

Reply:
Yeah that's fair, if they are useful cases then it makes sense to keep them. I think a different file would result in nicer output if there is a test failure, as it would tell you all the cases that fail rather than having to sort that out after the fact. Also, the simplest code would fail fast, and not failing fast results in a lot of bookkeeping code that is kind of what pytest is built to do. So I think moving it to a separate file is the way to go, at least for these 128 case ones.

Reply:
Yeah, a different file sounds good to me.